lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -119,121 +120,120 @@ def initialize(options = {}, &block)
         attr_reader :pool_size
 
         # Retrieves a connection. If there are active connections in the
-        # queue, the most recently used connection is returned. Otherwise
+        # stack, the most recently used connection is returned. Otherwise
         # if the connection pool size is less than the max size, creates a
         # new connection and returns it. Otherwise raises Timeout::Error.
         #
-        # @example Dequeue a connection.
-        #   queue.dequeue
+        # @example Pop a connection.
+        #   stack.pop
         #
         # @return [ Mongo::Server::Connection ] The next connection.
         # @raise [ Timeout::Error ] If the connection pool is at maximum size
         #   and remains so for longer than the wait timeout.
         #
         # @since 2.0.0
-        def dequeue
-          check_count_invariants
-          mutex.synchronize do
-            dequeue_connection
-          end
+        def pop(deadline)
+          check_count_invariants(false)
+          pop_connection(deadline)
         ensure
-          check_count_invariants
+          check_count_invariants(false)
         end
 
-        # Disconnect all connections in the queue.
+        # Updates the generation number. The connections will be disconnected and removed lazily
+        # when the stack attempts to pop them.
+        #
+        # @since 2.7.0
+        def clear!
+          @generation += 1
+        end
+
+        # Disconnect all connections in the stack.
         #
         # @example Disconnect all connections.
-        #   queue.disconnect!
+        #   stack.disconnect!
         #
         # @return [ true ] Always true.
         #
         # @since 2.1.0
         def disconnect!
-          check_count_invariants
+          check_count_invariants(false)
           mutex.synchronize do
-            queue.each{ |connection| connection.disconnect! }
-            @pool_size -= queue.length
-            if @pool_size < 0
-              # This should never happen
-              log_warn("ConnectionPool::Queue: connection accounting problem")
-              @pool_size = 0
+            connections.each do |connection|
+              connection_removed
+              connection.disconnect!
+
+              publish_cmap_event(
+                  Monitoring::Event::ConnectionClosed.new(
+                      Monitoring::Event::ConnectionClosed::POOL_CLOSED,
+                      @address,
+                      connection.id,
+                  ),
+              )
             end
-            queue.clear
+
+            connections.clear
             @generation += 1
```

Comment:
This generation bump should probably also happen prior to closing any of the connections.